File: go/timing/timing_test.go
// Timestamps: exact at the chunk, estimated at the word.
//
// The whole value of this feature is that a reading app can trust the first
// tier and be told, loudly, not to trust the second in the same way. So the
// tests are split the same way: the chunk assertions are equalities — down to
// the last bit, because a tolerance would hide exactly the defect that matters
// — and the word assertions are invariants (monotonic, inside the chunk, every
// word present). Nothing here claims a word lands where a listener would say it
// does.

package timing

import (
	"math"
	"testing"
)

const sampleRate = 24_000

// A highlight that switches on time >= start flickers on a gap and double-lights
// on an overlap, and both are invisible to a comparison with a tolerance.
// Offsets accumulate as integer samples for exactly this reason.
func TestChunksAreAdjacentToTheLastBit(t *testing.T) {
	got := Timeline([]Span{{"a b", 7_001, 3}, {"c d e", 13_337, 5}}, sampleRate)
	if got[0].Start != 0.0 {
		t.Errorf("the first chunk starts at %v, want 0", got[0].Start)
	}
	if got[1].Start != got[0].End {
		t.Errorf("chunk 1 starts at %v but chunk 0 ends at %v — a gap of %v",
			got[1].Start, got[0].End, got[1].Start-got[0].End)
	}
	if want := float64(7_001+13_337) / sampleRate; got[1].End != want {
		t.Errorf("the timeline ends at %v, want %v", got[1].End, want)
	}
}

func TestTheSpansCoverTheWholeRenderWithNothingLeftOver(t *testing.T) {
	got := Timeline([]Span{{"one", 100, 1}, {"two", 200, 2}, {"three", 300, 3}}, sampleRate)
	total := 0.0
	for _, c := range got {
		total += c.Duration()
	}
	if want := 600.0 / sampleRate; math.Abs(total-want) > 1e-12 {
		t.Errorf("the spans total %v seconds, want %v", total, want)
	}
	for i, want := range []int{1, 2, 3} {
		if got[i].Tokens != want {
			t.Errorf("chunk %d reports %d tokens, want %d", i, got[i].Tokens, want)
		}
	}
}

func TestAnEmptyRenderIsAnEmptyTimeline(t *testing.T) {
	if got := Timeline(nil, sampleRate); len(got) != 0 {
		t.Errorf("no spans gave %d timings", len(got))
	}
}

func TestWordsTileTheChunkWithoutGaps(t *testing.T) {
	words := EstimateWords("alpha beta gamma", 1.0, 4.0)
	if len(words) != 3 {
		t.Fatalf("got %d words, want 3", len(words))
	}
	if words[0].Start != 1.0 {
		t.Errorf("the first word starts at %v, want the chunk's own start", words[0].Start)
	}
	if words[2].End != 4.0 {
		t.Errorf("the last word ends at %v, want the chunk's own end", words[2].End)
	}
	for i := 0; i+1 < len(words); i++ {
		if words[i].End != words[i+1].Start {
			t.Errorf("word %d ends at %v but word %d starts at %v",
				i, words[i].End, i+1, words[i+1].Start)
		}
	}
}

func TestTimesAreMonotonicAndInsideTheChunk(t *testing.T) {
	words := EstimateWords("a bb ccc dddd e", 2.5, 3.25)
	previous := 2.5
	for _, w := range words {
		if !(2.5 <= w.Start && w.Start <= w.End && w.End <= 3.25) {
			t.Errorf("%q spans [%v, %v], outside the chunk's [2.5, 3.25]", w.Text, w.Start, w.End)
		}
		if w.Start < previous {
			t.Errorf("%q starts at %v, before the previous word's start %v", w.Text, w.Start, previous)
		}
		previous = w.Start
	}
}

// The whole content of the estimate: characters stand in for seconds. Nothing
// else here knows how long a word takes.
func TestALongerWordIsGivenLonger(t *testing.T) {
	words := EstimateWords("hi internationalisation", 0.0, 1.0)
	if len(words) != 2 {
		t.Fatalf("got %d words, want 2", len(words))
	}
	short := words[0].End - words[0].Start
	long := words[1].End - words[1].Start
	if long <= short {
		t.Errorf("the long word got %v s and the short one %v s", long, short)
	}
}

// A caller highlighting "end." wants the full stop lit with the word, and a
// caller matching back against their own text needs the substring to be a
// substring.
func TestPunctuationStaysWithItsWord(t *testing.T) {
	words := EstimateWords("Hello, world!", 0.0, 1.0)
	if len(words) != 2 || words[0].Text != "Hello," || words[1].Text != "world!" {
		t.Errorf("got %v, want [Hello, world!] with the punctuation attached", words)
	}
}

func TestNoTextIsNoWordsRatherThanADivisionByZero(t *testing.T) {
	if got := EstimateWords("   ", 0.0, 1.0); len(got) != 0 {
		t.Errorf("whitespace gave %d words", len(got))
	}
	if got := EstimateWords("", 0.0, 1.0); len(got) != 0 {
		t.Errorf("an empty string gave %d words", len(got))
	}
}

// The other four ports count code points too. A byte count would give Polish and
// Japanese text different word weights in Go than in Python, for text that reads
// identically — "żółć" is four characters to a reader and eight bytes to Go.
func TestLengthIsCountedInCodePointsNotBytes(t *testing.T) {
	words := EstimateWords("aaaa żółć", 0.0, 1.0)
	if len(words) != 2 {
		t.Fatalf("got %d words, want 2", len(words))
	}
	ascii := words[0].End - words[0].Start
	accented := words[1].End - words[1].Start
	if math.Abs(ascii-accented) > 1e-12 {
		t.Errorf("four ASCII characters got %v s and four accented ones %v s", ascii, accented)
	}
}

func TestEstimateWordsSurvivesAChunkOfOneWord(t *testing.T) {
	words := EstimateWords("word", 0.5, 0.75)
	if len(words) != 1 || words[0].Start != 0.5 || words[0].End != 0.75 {
		t.Errorf("got %v, want one word spanning exactly [0.5, 0.75]", words)
	}
}

// What a caller stitching a stream does with each chunk's own timing.
func TestShiftingMovesTheWordsWithTheChunk(t *testing.T) {
	span := Timeline([]Span{{"a bb", 240, 2}}, sampleRate)[0]
	moved := span.Shifted(1.0)
	if moved.Start != span.Start+1.0 || moved.End != span.End+1.0 {
		t.Errorf("the chunk moved to [%v, %v], want [%v, %v]",
			moved.Start, moved.End, span.Start+1.0, span.End+1.0)
	}
	if len(moved.Words) != len(span.Words) {
		t.Fatalf("shifting changed the word count: %d -> %d", len(span.Words), len(moved.Words))
	}
	for i, w := range moved.Words {
		if w.Start != span.Words[i].Start+1.0 || w.Text != span.Words[i].Text {
			t.Errorf("word %d moved to %v (%q), want %v (%q)",
				i, w.Start, w.Text, span.Words[i].Start+1.0, span.Words[i].Text)
		}
	}
	// The original is untouched: Shifted returns a new timing rather than
	// editing the words in place, and a caller stitching a stream keeps both.
	if span.Start != 0.0 {
		t.Errorf("shifting moved the original too: it now starts at %v", span.Start)
	}
}

File: js/src/test/timing.test.ts
/**
 * The two timing tiers: exact chunk spans, estimated word spans.
 *
 * No fixture and no weights — the arithmetic is the whole feature, and it is
 * pure. What is pinned here is what a highlight in a reading app depends on:
 * that the joins between chunks are the *same float* on both sides (not merely
 * close), that the words tile their chunk without drifting, and that a word's
 * weight is its code-point count so the five ports weight the same text the same
 * way.
 */

import assert from "node:assert/strict";
import { test } from "node:test";

import { estimateWords, timeline } from "../timing.js";

const RATE = 24_000;

test("chunk joins are the same float on both sides, not merely close", () => {
  // Sample counts chosen so none of the offsets is exactly representable as a
  // short decimal: a timeline that accumulated seconds would differ here in the
  // last bit, which is a nanosecond gap — invisible to an assertion with a
  // tolerance, visible as a flicker in a highlight that switches on
  // `time >= start`.
  const spans = [
    { text: "One.", samples: 12_345, tokens: 41 },
    { text: "Two.", samples: 7, tokens: 1 },
    { text: "Three.", samples: 98_765, tokens: 190 },
  ];
  const chunks = timeline(spans, RATE);

  assert.equal(chunks.length, 3);
  assert.equal(chunks[0].start, 0);
  for (let i = 0; i + 1 < chunks.length; i++) {
    assert.equal(chunks[i].end, chunks[i + 1].start, `join ${i} is not exact`);
  }
  const total = spans.reduce((n, s) => n + s.samples, 0);
  assert.equal(chunks[chunks.length - 1].end, total / RATE);
  // Text and token counts travel with the span rather than being recomputed.
  assert.deepEqual(
    chunks.map((c) => [c.text, c.tokens]),
    [
      ["One.", 41],
      ["Two.", 1],
      ["Three.", 190],
    ]
  );
});

test("one span covers the whole render", () => {
  const [only] = timeline([{ text: "Hello there.", samples: RATE * 2, tokens: 50 }], RATE);
  assert.equal(only.start, 0);
  assert.equal(only.end, 2);
  assert.equal(only.words.length, 2);
});

test("word spans tile their chunk, monotonic and bounded", () => {
  const [chunk] = timeline([{ text: "one two three four", samples: RATE, tokens: 25 }], RATE);
  const words = chunk.words;
  assert.deepEqual(
    words.map((w) => w.text),
    ["one", "two", "three", "four"]
  );
  // The first start and the last end are the chunk's own, exactly: the
  // boundaries come from a running character total rather than from summing
  // per-word durations, so they cannot drift.
  assert.equal(words[0].start, chunk.start);
  assert.equal(words[words.length - 1].end, chunk.end);
  for (let i = 0; i < words.length; i++) {
    assert.ok(words[i].end >= words[i].start, `word ${i} runs backwards`);
    assert.ok(words[i].start >= chunk.start && words[i].end <= chunk.end, `word ${i} escapes`);
    if (i + 1 < words.length) {
      assert.equal(words[i].end, words[i + 1].start, `word join ${i} is not shared`);
    }
  }
});

test("a word's weight is its code-point count, not its UTF-16 length", () => {
  // "𝐚𝐛" is two code points and four UTF-16 units. Weighing it by `.length`
  // would give it two thirds of the span here while Python, Go, Rust and Swift
  // gave it half — the same text read as two different timings, with nothing to
  // show for it in any output anyone looks at.
  const [chunk] = timeline([{ text: "\u{1D41A}\u{1D41B} cd", samples: RATE, tokens: 25 }], RATE);
  assert.equal(chunk.words.length, 2);
  assert.equal(chunk.words[0].end, 0.5);
  assert.equal(chunk.words[1].start, 0.5);
});

test("punctuation stays attached to the word it follows", () => {
  // The split is on whitespace on purpose: a caller lighting up the word wants
  // the full stop lit with it, and a caller matching back against their own
  // string needs the word to be a substring of it.
  const words = estimateWords("Hello, world!", 0, 1);
  assert.deepEqual(
    words.map((w) => w.text),
    ["Hello,", "world!"]
  );
});

test("text with no words yields no words", () => {
  assert.deepEqual(estimateWords("", 0, 1), []);
  assert.deepEqual(estimateWords("   \n\t ", 0, 1), []);
  // Leading and trailing whitespace is dropped rather than becoming an empty
  // word with a zero-length span, which is what a naive split on " " gives.
  assert.deepEqual(
    estimateWords("  spaced   out  ", 0, 1).map((w) => w.text),
    ["spaced", "out"]
  );
});

test("an empty timeline is empty, not one zero-length chunk", () => {
  assert.deepEqual(timeline([], RATE), []);
});
